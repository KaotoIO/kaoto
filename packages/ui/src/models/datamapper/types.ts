/**
 * The data types mapped from XML Schema +.
 */
export enum Types {
  AnyAtomicType = 'anyAtomicType',
  AnyType = 'anyType',
  AnyURI = 'anyURI',
  String = 'string',
  Float = 'float',
  Double = 'double',
  Integer = 'integer',
  Decimal = 'decimal',
  Boolean = 'boolean',
  Date = 'date',
  DateTime = 'dateTime',
  Time = 'time',
  QName = 'QName',
  Duration = 'duration',
  DayTimeDuration = 'dayTimeDuration',
  NCName = 'NCName',
  // not XSD predefined but appear in XPath function signature
  Numeric = 'numeric',
  Item = 'item()',
  Element = 'element()',
  Node = 'node()',
  DocumentNode = 'document-node()',
  // custom types
  Container = 'Container',
  Array = 'Array',
}
