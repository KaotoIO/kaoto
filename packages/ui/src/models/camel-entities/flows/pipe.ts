import { Pipe as PipeModel } from '@kaoto-next/camel-catalog/types';
import { v4 as uuidv4 } from 'uuid';
import { BaseVisualCamelEntity, EntityType } from '../base-entity';
import { KameletBinding } from './kamelet-binding';

export class Pipe extends KameletBinding implements BaseVisualCamelEntity {
  readonly id = uuidv4();
  type = EntityType.Pipe;

  constructor(public route: Partial<PipeModel> = {}) {
    super();
  }
}
