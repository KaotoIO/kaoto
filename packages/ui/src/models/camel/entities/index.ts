export * from './base-entity';
export * from './kamelet-binding-overrides';
export * from './pipe-overrides';
