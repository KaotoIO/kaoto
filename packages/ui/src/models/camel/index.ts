export * from './camel-resource';
export * from './camel-route-resource';
export * from './integration-resource';
export * from './kamelet-resource';
export * from './kamelet-binding-resource';
export * from './pipe-resource';
export * from './source-schema-config';
export * from './source-schema-type';
export { CAMEL_K_K8S_API_VERSION_V1 } from './camel-k-resource';
