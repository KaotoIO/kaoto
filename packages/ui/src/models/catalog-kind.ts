export const enum CatalogKind {
  Component = 'component',
  Processor = 'processor',
  Pattern = 'pattern',
  Language = 'language',
  Dataformat = 'dataformat',
  Kamelet = 'kamelet',
}
