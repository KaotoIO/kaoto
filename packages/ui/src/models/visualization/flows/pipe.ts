import { Pipe as PipeModel } from '@kaoto-next/camel-catalog/types';
import { v4 as uuidv4 } from 'uuid';
import { BaseVisualCamelEntity } from '../base-visual-entity';
import { EntityType } from '../../camel-entities/base-entity';
import { KameletBinding } from './kamelet-binding';

export class Pipe extends KameletBinding implements BaseVisualCamelEntity {
  readonly id = uuidv4();
  type = EntityType.Pipe;

  constructor(public route: Partial<PipeModel> = {}) {
    super();
  }
}
