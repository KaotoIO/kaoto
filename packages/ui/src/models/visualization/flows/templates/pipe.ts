import { getCamelRandomId } from '../../../../camel-utils/camel-random-id';

export const pipeTemplate = () => {
  return `apiVersion: camel.apache.org/v1
kind: Pipe
metadata:
  name: ${getCamelRandomId('pipe')}
spec:
  source:
    properties:
      message: hello
    ref:
      kind: Kamelet
      apiVersion: camel.apache.org/v1
      name: timer-source
  sink:
    ref:
      kind: Kamelet
      apiVersion: camel.apache.org/v1
      name: log-sink`;
};
