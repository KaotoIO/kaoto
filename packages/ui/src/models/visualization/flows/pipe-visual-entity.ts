import { Pipe } from '@kaoto-next/camel-catalog/types';
import get from 'lodash.get';
import set from 'lodash.set';
import { getCamelRandomId } from '../../../camel-utils/camel-random-id';
import { getArrayProperty } from '../../../utils';
import { NodeIconResolver } from '../../../utils/node-icon-resolver';
import { DefinedComponent } from '../../camel-catalog-index';
import { EntityType } from '../../camel/entities';
import { PipeMetadata, PipeSpec, PipeStep, PipeSteps } from '../../camel/entities/pipe-overrides';
import {
  AddStepMode,
  BaseVisualCamelEntity,
  IVisualizationNode,
  IVisualizationNodeData,
  NodeInteraction,
  VisualComponentSchema,
} from '../base-visual-entity';
import { createVisualizationNode } from '../visualization-node';
import { KameletSchemaService } from './support/kamelet-schema.service';

export class PipeVisualEntity implements BaseVisualCamelEntity {
  id: string;
  type = EntityType.Pipe;
  spec: PipeSpec;
  metadata: PipeMetadata;

  constructor(spec?: PipeSpec, metadata?: PipeMetadata) {
    this.id = (metadata?.name as string) ?? getCamelRandomId('pipe');
    this.metadata = metadata ?? { name: this.id };
    this.spec = spec ?? {
      source: {},
      steps: [],
      sink: {},
    };
  }

  /** Internal API methods */
  getId(): string {
    return this.id;
  }

  setId(routeId: string): void {
    this.id = routeId;
    this.metadata.name = this.id;
  }

  getNodeLabel(path?: string): string {
    if (!path) return '';

    const stepModel = get(this.spec, path) as PipeStep;
    return KameletSchemaService.getNodeLabel(stepModel, path);
  }

  getComponentSchema(path?: string): VisualComponentSchema | undefined {
    if (!path) return undefined;
    const stepModel = get(this.spec, path) as PipeStep;
    return KameletSchemaService.getVisualComponentSchema(stepModel);
  }

  toJSON() {
    return this.spec;
  }

  updateModel(path: string | undefined, value: unknown): void {
    if (!path) return;

    const stepModel = get(this.spec, path) as PipeStep;
    if (stepModel) set(stepModel, 'properties', value);
  }

  getSteps() {
    const steps: PipeSteps = this.spec?.steps;
    const sink: PipeStep = this.spec?.sink;
    let allSteps: Array<PipeStep> = [];
    if (steps !== undefined) {
      allSteps = allSteps.concat(steps);
    }
    !sink || Object.keys(sink).length === 0 || allSteps.push(sink);

    return allSteps;
  }

  /**
   * Add a step to the Pipe
   *
   * path examples:
   *      source
   *      sink
   *      steps.0
   *      steps.1
   */
  addStep(options: {
    definedComponent: DefinedComponent;
    mode: AddStepMode;
    data: IVisualizationNodeData;
    targetProperty?: string;
  }) {
    const newKamelet = options.definedComponent.definition as unknown as Pipe;
    const path = options.data.path;
    if (!newKamelet || !path) return;

    const step: PipeStep = {
      ref: {
        kind: newKamelet.kind,
        apiVersion: newKamelet.apiVersion,
        name: options.definedComponent.name,
      },
    };

    /** Replace an existing Kamelet */
    if (options.mode === AddStepMode.ReplaceStep) {
      if (path === 'source' || path === 'sink') {
        set(this.spec, path, step);
      } else {
        set(this.spec, path, step);
      }

      return;
    }

    /** Add a new Kamelet to the Kamelets array */
    const kameletArray = getArrayProperty(this.spec, 'steps') as PipeStep[];
    const index = Number(path.split('.').pop());
    if (options.mode === AddStepMode.AppendStep) {
      kameletArray.splice(index + 1, 0, step);
    } else if (options.mode === AddStepMode.PrependStep && options.data.path === 'sink') {
      kameletArray.push(step);
    } else if (options.mode === AddStepMode.PrependStep) {
      kameletArray.splice(index, 0, step);
    }
  }

  removeStep(path?: string): void {
    /** This method needs to be enabled after passing the entire parent to this class*/
    if (!path) return;
    /**
     * If the path is `source` or `sink`, we can remove it directly
     */
    if (path === 'source' || path === 'sink') {
      set(this.spec, path, {});
      return;
    }
    const pathArray = path.split('.');
    const last = pathArray[pathArray.length - 1];
    /**
     * If the last segment is a number, it means the target object is a member of an array
     * therefore we need to look for the array and remove the element at the given index
     *
     * f.i. from.steps.1.choice.when.0
     * last: 0
     */
    const array = get(this.spec, pathArray.slice(0, -1), []);
    if (Number.isInteger(Number(last)) && Array.isArray(array)) {
      array.splice(Number(last), 1);
    }
  }

  getNodeInteraction(data: IVisualizationNodeData): NodeInteraction {
    return {
      /** Pipe cannot have a Kamelet before the source property */
      canHavePreviousStep: data.path !== 'source',
      /** Pipe cannot have a Kamelet after the sink property */
      canHaveNextStep: data.path !== 'sink',
      canHaveChildren: false,
      canHaveSpecialChildren: false,
    };
  }

  getNodeValidationText(_path?: string | undefined): string | undefined {
    return undefined;
  }

  toVizNode(): IVisualizationNode {
    const sourceNode = this.getVizNodeFromStep(this.spec.source, 'source', true);
    const stepNodes = this.getVizNodesFromSteps(this.spec.steps);
    const sinkNode = this.getVizNodeFromStep(this.spec.sink, 'sink');
    /** If there are no steps, we link the `source` and the `sink` together */

    if (stepNodes.length === 0) {
      sourceNode.setNextNode(sinkNode);
      sinkNode.setPreviousNode(sourceNode);
      return sourceNode;
    }

    /** Connect the `source` with the first step */
    const firstStepNode = stepNodes[0];
    if (firstStepNode !== undefined) {
      sourceNode.setNextNode(firstStepNode);
      firstStepNode.setPreviousNode(sourceNode);
    }

    /** Connect the last step with the `sink` */
    const lastStepNode = stepNodes[stepNodes.length - 1];
    if (lastStepNode !== undefined) {
      lastStepNode.setNextNode(sinkNode);
      sinkNode.setPreviousNode(lastStepNode);
    }

    return sourceNode;
  }

  private getVizNodeFromStep(step: PipeStep, path: string, isRoot = false): IVisualizationNode {
    const kameletDefinition = KameletSchemaService.getKameletDefinition(step);
    const isPlaceholder = step?.ref?.name === undefined;
    const icon = isPlaceholder
      ? NodeIconResolver.getPlaceholderIcon()
      : kameletDefinition?.metadata.annotations['camel.apache.org/kamelet.icon'] ?? NodeIconResolver.getUnknownIcon();

    const data: IVisualizationNodeData = {
      path,
      entity: isRoot ? this : undefined,
      isPlaceholder,
      icon,
    };

    return createVisualizationNode(step?.ref?.name ?? path, data);
  }

  private getVizNodesFromSteps(steps?: PipeStep[]): IVisualizationNode[] {
    if (!Array.isArray(steps)) {
      return [] as IVisualizationNode[];
    }

    return steps?.reduce((acc, kamelet) => {
      const vizNode = this.getVizNodeFromStep(kamelet, 'steps.' + acc.length);

      const previousVizNode = acc[acc.length - 1];
      if (previousVizNode !== undefined) {
        previousVizNode.setNextNode(vizNode);
        vizNode.setPreviousNode(previousVizNode);
      }

      acc.push(vizNode);
      return acc;
    }, [] as IVisualizationNode[]);
  }
}
