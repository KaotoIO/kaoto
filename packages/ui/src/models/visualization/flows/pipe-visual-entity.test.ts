import * as catalogIndex from '@kaoto-next/camel-catalog/index.json';
import { Pipe } from '@kaoto-next/camel-catalog/types';
import cloneDeep from 'lodash/cloneDeep';
import { pipeJson } from '../../../stubs/pipe';
import { EntityType } from '../../camel/entities';
import { CatalogKind } from '../../catalog-kind';
import { IKameletDefinition } from '../../kamelets-catalog';
import { KaotoSchemaDefinition } from '../../kaoto-schema';
import { CamelCatalogService } from './camel-catalog.service';
import { PipeVisualEntity } from './pipe-visual-entity';
import { KameletSchemaService } from './support/kamelet-schema.service';

describe('Pipe', () => {
  let pipeCR: Pipe;
  let pipe: PipeVisualEntity;
  let kameletCatalogMap: Record<string, unknown>;

  beforeEach(async () => {
    pipeCR = cloneDeep(pipeJson);
    pipe = new PipeVisualEntity(pipeCR.spec!);
    kameletCatalogMap = await import('@kaoto-next/camel-catalog/' + catalogIndex.catalogs.kamelets.file);
    CamelCatalogService.setCatalogKey(CatalogKind.Kamelet, kameletCatalogMap as Record<string, IKameletDefinition>);
  });

  describe('id', () => {
    it('should have an uuid', () => {
      expect(pipe.id).toBeDefined();
      expect(typeof pipe.id).toBe('string');
    });

    it('should have a type', () => {
      expect(pipe.type).toEqual(EntityType.Pipe);
    });

    it('should return the id', () => {
      expect(pipe.getId()).toEqual(expect.any(String));
    });

    it('should change the id', () => {
      pipe.setId('pipe-12345');
      expect(pipe.getId()).toEqual('pipe-12345');
    });
  });

  describe('getComponentSchema', () => {
    it('should return undefined if no path is provided', () => {
      expect(pipe.getComponentSchema()).toBeUndefined();
    });

    it('should return undefined if no component model is found', () => {
      const result = pipe.getComponentSchema('test');

      expect(result).toBeUndefined();
    });

    it('should return the component schema', () => {
      const spy = jest.spyOn(KameletSchemaService, 'getVisualComponentSchema');
      spy.mockReturnValueOnce({
        title: 'test',
        schema: {} as KaotoSchemaDefinition['schema'],
        definition: {},
      });

      pipe.getComponentSchema('source');
      expect(spy).toBeCalledTimes(1);
    });
  });

  it('should return the json', () => {
    expect(pipe.toJSON()).toEqual(pipeJson.spec);
  });

  describe('updateModel', () => {
    it('should not update the model if no path is provided', () => {
      const originalObject = JSON.parse(JSON.stringify(pipeJson));

      pipe.updateModel(undefined, undefined);

      expect(originalObject).toEqual(pipeJson);
    });

    it('should update the model', () => {
      const name = 'webhook-source';

      pipe.updateModel('source', { name });

      expect(pipe.spec?.source?.properties?.name).toEqual(name);
    });

    it('should not update the model if the path is not correct', () => {
      const name = 'webhook-source';

      pipe.updateModel('nonExistingPath', { name });

      expect(pipe.spec?.source?.properties).toBeUndefined();
    });
  });

  describe('removeStep', () => {
    it('should not remove the step if no path is provided', () => {
      pipe.removeStep();

      expect(pipe.toJSON()).toEqual(pipeJson.spec);
    });

    it('should remove the `source` step', () => {
      pipe.removeStep('source');

      expect(pipe.toJSON()).not.toEqual(pipeJson.spec);
      expect(pipe.spec?.source).toEqual({});
    });

    it('should remove the `sink` step', () => {
      pipe.removeStep('sink');

      expect(pipe.toJSON()).not.toEqual(pipeJson.spec);
      expect(pipe.spec?.sink).toEqual({});
    });

    it('should remove the `steps.0` step', () => {
      pipe.removeStep('steps.0');

      expect(pipe.toJSON()).not.toEqual(pipeJson.spec);
      expect(pipe.spec.steps).toEqual([]);
    });
  });

  describe('getNodeInteraction', () => {
    it.each(['source', 'sink', 'steps.1', '#'])(
      `should return the correct interaction for the '%s' processor`,
      (path) => {
        const result = pipe.getNodeInteraction({ path });
        expect(result).toMatchSnapshot();
      },
    );
  });

  describe('getNodeValidationText', () => {
    it('should return an `undefined` if the path is `undefined`', () => {
      const result = pipe.getNodeValidationText(undefined);

      expect(result).toEqual(undefined);
    });

    it('should return an `undefined` if the path is empty', () => {
      const result = pipe.getNodeValidationText('');

      expect(result).toEqual(undefined);
    });

    it('should return a validation text relying on the `validateNodeStatus` method', () => {
      const missingParametersModel = cloneDeep(pipeJson.spec);
      missingParametersModel!.steps![0].properties = {};
      pipe = new PipeVisualEntity(missingParametersModel);

      const result = pipe.getNodeValidationText('steps.0');

      expect(result).toEqual('1 required parameter is not yet configured: [ milliseconds ]');
    });
  });

  describe('toVizNode', () => {
    it('should return the viz node and set the initial path to `source`', () => {
      const vizNode = pipe.toVizNode();

      expect(vizNode).toBeDefined();
      expect(vizNode.data.path).toEqual('source');
    });

    it('should use the uri as the node label', () => {
      const vizNode = pipe.toVizNode();

      expect(vizNode.getNodeLabel()).toEqual('webhook-source');
    });

    it('should set the node labels as `Unknown` if the uri is not available', () => {
      pipe = new PipeVisualEntity({});
      const sourceNode = pipe.toVizNode();
      const sinkNode = sourceNode.getNextNode();

      expect(sourceNode.getNodeLabel()).toEqual('source: Unknown');
      expect(sinkNode!.getNodeLabel()).toEqual('sink: Unknown');
    });

    it('should populate the viz node chain with the steps', () => {
      const vizNode = pipe.toVizNode();

      expect(vizNode.data.path).toEqual('source');
      expect(vizNode.getNodeLabel()).toEqual('webhook-source');
      expect(vizNode.getPreviousNode()).toBeUndefined();
      expect(vizNode.getNextNode()).toBeDefined();

      const steps0 = vizNode.getNextNode()!;
      expect(steps0.data.path).toEqual('steps.0');
      expect(steps0.getNodeLabel()).toEqual('delay-action');
      expect(steps0.getPreviousNode()).toBe(vizNode);
      expect(steps0.getNextNode()).toBeDefined();

      const sink = steps0.getNextNode()!;
      expect(sink.data.path).toEqual('sink');
      expect(sink.getNodeLabel()).toEqual('log-sink');
      expect(sink.getPreviousNode()).toBe(steps0);
      expect(sink.getNextNode()).toBeUndefined();
    });
  });
});
