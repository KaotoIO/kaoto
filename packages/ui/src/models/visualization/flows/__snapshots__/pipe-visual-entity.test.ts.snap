// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipe getNodeInteraction should return the correct interaction for the '#' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": false,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": true,
  "canRemoveStep": false,
  "canReplaceStep": false,
}
`;

exports[`Pipe getNodeInteraction should return the correct interaction for the 'sink' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": false,
  "canHavePreviousStep": true,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`Pipe getNodeInteraction should return the correct interaction for the 'source' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": true,
  "canHavePreviousStep": false,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;

exports[`Pipe getNodeInteraction should return the correct interaction for the 'steps.1' processor 1`] = `
{
  "canBeDisabled": false,
  "canHaveChildren": false,
  "canHaveNextStep": true,
  "canHavePreviousStep": true,
  "canHaveSpecialChildren": false,
  "canRemoveFlow": false,
  "canRemoveStep": true,
  "canReplaceStep": true,
}
`;
